[14.0][MIG] product_state_history

rousseldenis opened the pull request:
Depends on:

- [x] #725

rousseldenis committed:
[10.0][ADD] product_state_history

rousseldenis committed:
[10.0][IMP] product_state_history: Get good state

Take into account product template status instead of history
one when launching report

rousseldenis committed:
[10.0][IMP] product_state_history: Change behaviour

Wrong assertion for report as it should include products AFTER
pivot date.

Better view definition

rousseldenis committed:
[10.0][IMP] product_state_history: Better name

To distinguish the report and the history (all the records),
the action name has been changed

oca-travis committed:
[UPD] Update product_state_history.pot

OCA-git-bot committed:
product_state_history 10.0.1.0.1

OCA-git-bot committed:
[UPD] README.rst

rousseldenis committed:
[IMP] product_state_history: black, isort, prettier

rousseldenis committed:
[14.0][MIG] product_state_history

hailangvn reviewed (approved):
Code review: LGTM. Thank you.

xavier-bouquiaux commented on product_state_history/models/product_template.py:
```diff
@@ -0,0 +1,42 @@
+        history_vals = []
+        state_id = vals["product_state_id"]
+        for template in self:
+            history_vals.append(
```
@rousseldenis 
Line are duplicated for now (the change in history are shown 2 times)
in debug we saw that the write is called 2 times when we change a product_state
could you add this if please ?

```
     if state_id != template.product_state_id.id:
                history_vals.append(
                    template._prepare_product_state_history_values(state_id)
                )
```

rousseldenis replied:
@xavier-bouquiaux @sbejaoui I agree with this protection.

In fact, it should have been a good point to check why two write() are called.

This is due to product_state implementation as both state and state_id are in the view.

I will add a readonly=True to state field there https://github.com/OCA/product-attribute/blob/14.0/product_state/models/product_template.py#L15

rousseldenis committed:
[14.0][IMP] product_state_history: Prevent a duplicate history

If the same state than the current is written, a duplicate state history
is generated. This prevents that behaviour.

sbejaoui commented on product_state_history/models/product_template.py:
```diff
@@ -0,0 +1,43 @@
+        history_vals = []
+        state_id = vals["product_state_id"]
+        # We check if new state is different from the current one
+        for template in self.filtered(lambda t: t.product_state_id.id != state_id):
```
it's better to add the if inside the for loop, otherwise it will loop twice

```suggestion
        for template in self:
            if template.product_state_id.id != state_id:
```

dreispt commented:
/ocabot merge nobump
